enterprise/suppress-user/handler.go

```diff
@@ -0,0 +1,51 @@
+package suppression
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/samber/lo"
+)
+
+// newHandler creates a new handler for the suppression feature
+func newHandler(ctx context.Context, r Repository, log logger.Logger) *handler {
+	h := &handler{
+		r:   r,
+		log: log,
+	}
+	// we don't want to flood logs if, e.g. the suppression repository is restoring,
+	// so we are debouncing the logging
+	h.errLog.debounceLog, h.errLog.cancel = lo.NewDebounce(1*time.Second, func() {
+		if h.errLog.lastError != nil {
```

Comment:
This function is still invoked even if once every second so if we call `IsSuppressedUser` we have a data race on `lastError`.

```
==================
WARNING: DATA RACE
Write at 0x00c0001ecaf0 by goroutine 38:
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.(*handler).IsSuppressedUser()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler.go:47 +0x371
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.TestConcurrency.func1()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler_test.go:23 +0xc4

Previous write at 0x00c0001ecaf0 by goroutine 44:
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.(*handler).IsSuppressedUser()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler.go:47 +0x371
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.TestConcurrency.func1()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler_test.go:23 +0xc4

Goroutine 38 (running) created at:
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.TestConcurrency()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler_test.go:21 +0xbc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47

Goroutine 44 (running) created at:
  github.com/rudderlabs/rudder-server/enterprise/suppress-user.TestConcurrency()
      /home/francesco/Code/rudderstack/rudder-server/enterprise/suppress-user/handler_test.go:21 +0xbc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1446 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1493 +0x47
==================
    testing.go:1319: race detected during execution of test
--- FAIL: TestConcurrency (0.01s)
=== CONT  
    testing.go:1319: race detected during execution of test
FAIL
FAIL	github.com/rudderlabs/rudder-server/enterprise/suppress-user	0.034s
FAIL
```

### handler_test.go

```go
package suppression

import (
	"context"
	"fmt"
	"io"
	"sync"
	"testing"

	"github.com/rudderlabs/rudder-server/enterprise/suppress-user/model"
	"github.com/rudderlabs/rudder-server/utils/logger"
)

func TestConcurrency(t *testing.T) {
	ctx := context.Background()
	h := newHandler(ctx, &fakeSuppresser{}, logger.NOP)

	var wg sync.WaitGroup
	wg.Add(1000)
	for i := 0; i < 1000; i++ {
		go func() {
			defer wg.Done()
			_ = h.IsSuppressedUser("workspaceID", "userID", "sourceID")
		}()
	}

	wg.Wait()
}

type fakeSuppresser struct{}

func (*fakeSuppresser) Stop() error                               { return nil }
func (*fakeSuppresser) GetToken() ([]byte, error)                 { return nil, nil }
func (*fakeSuppresser) Add(_ []model.Suppression, _ []byte) error { return nil }
func (*fakeSuppresser) Backup(_ io.Writer) error                  { return nil }
func (*fakeSuppresser) Restore(_ io.Reader) error                 { return nil }
func (*fakeSuppresser) Suppress(_, _, _ string) (bool, error) {
	return false, fmt.Errorf("some error")
}
```

Reply:
The purpose of the debouncer here0 is to avoid sending a flood of logs in case of errors (i.e. when the repository is still restoring). Just one of the errors recorded every second should be enough. I don't think we should care about a data race here, unless I am unintentionally ignoring something.

Reply:
I know in the C++ world people have developed with benign data races for a long time until some guy came up with a document explaning how benign data races can create issues (see [How to miscompile programs with “benign” data races](https://www.usenix.org/legacy/event/hotpar11/tech/final_files/Boehm.pdf). See also [here](https://bartoszmilewski.com/2020/08/11/benign-data-races-considered-harmful/).

I don't know enough of Golang internals to back this up here but compilers and memory models do change over time so my suggestion is to avoid data races, even if benign they could cause issues that might be hard to foresee.

Additionally our plan is to run all tests with `-race` in `make test`. In that case more coverage in this package would make the tests fail.

Reply:
I wouldn't really justify a mutex for protecting this error message, do you believe it is needed?

Reply:
If I think that this will make test with `-race` yes, definitely. I do not see any harm on adding it as well. Anyway, beside the test scenario, as I mentioned I do not know enough of Golang internals but I do know that compilers and memory models change over time so to stay on the safe side I'd do it the canonical way just in case.